### PR TITLE
add password property to CREATE USER statement

### PR DESCRIPTION
**Pull request by mikethebeer**
This will extend the CREATE USER statement with an optional password property. It allows to specify a password for the user that is being created.

Usage:

```sql
CREATE USER foo WITH (password = 'pass');
```

The password information is computed using a salted PBKDF2 algorithm that produces a SHA-512 hash. The `SecureHash` (containing iterations, salt and SHA-512 hash) is stored in the custom metdata.

To be followed:
* User Meta Data Migration layer that ensures correct user-data migration if Crate gets updated.

### CODE REVIEW

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 1]
```diff
@@ -30,87 +32,159 @@
...
+                        switch (currentFieldName) {
+                            case "iterations":
+                                if (currentToken != XContentParser.Token.VALUE_NUMBER) {
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'iterations' value is not a number [{}]", currentToken);
+                                }
+                                iterations = parser.intValue();
+                                break;
+                            case "hash":
+                                if (currentToken == XContentParser.Token.VALUE_NULL) {
```
Comment (mikethebeer) [inline enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: @seut do you have a proper solution what Token type for `byte[]` could be used instead of a `VALUE_NULL` check ?

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 2]
```diff
@@ -30,87 +32,159 @@
...
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'hash' value is not a binary value [{}]", currentToken);
+                                }
+                                hash = parser.binaryValue();
+                                break;
+                            case "salt":
+                                if (currentToken == XContentParser.Token.VALUE_NULL) {
```
Comment (mikethebeer) [inline enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: @seut same here. Checking the Token now with `VALUE_NULL` since there is not proper Token-Type that I could use for `byte[]` types

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 3]
```diff
@@ -30,76 +32,130 @@
...
+            token = parser.nextToken();
+            if (token == XContentParser.Token.START_ARRAY) {
+                // UsersMetaData v1
                 while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY && token != null) {
-                    users.add(parser.text());
+                    users.put(parser.text(), null); // old users do not have passwords
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                // UsersMetaData v2
+                while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
```
Comment (mxm) [inline enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: I thought we wanted to prevent such conditions in favor of a proper migration *before* reading any user data.

File: enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java [hunk 4]
```diff
@@ -130,4 +130,10 @@ public void testDropSuperUserThrowsException() throws Exception {
         expectedException.expectMessage("UnsupportedFeatureException: Cannot drop a superuser 'crate'");
         executeAsSuperuser("drop user crate");
     }
+
+    @Test
+    public void testCreateUserWithPassword() throws Exception {
+        executeAsSuperuser("create user foo with (password = 'password')");
+        assertUserIsCreated("foo");
```
Comment (mxm) [inline enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: Can we check the password as well?

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 5]
```diff
@@ -30,87 +32,159 @@
...
-                XContentParser.Token token;
-                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY && token != null) {
-                    users.add(parser.text());
-                }
-            }
-            if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
-                // each custom metadata is packed inside an object.
-                // each custom must move the parser to the end otherwise possible following customs won't be read
-                throw new ElasticsearchParseException("failed to parse users, expected an object token at the end");
+        Map<String, UserAttributes> users = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (mikethebeer) [reply enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: addressed here https://github.com/crate/crate/pull/6519/files#diff-b619b1318ed83f82c5912a603a130c5cL105

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 6]
```diff
@@ -30,87 +32,159 @@
...
+                                if (currentToken != XContentParser.Token.VALUE_NUMBER) {
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'iterations' value is not a number [{}]", currentToken);
+                                }
+                                iterations = parser.intValue();
+                                break;
+                            case "hash":
+                                if (currentToken == XContentParser.Token.VALUE_NULL) {
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'hash' value is not a binary value [{}]", currentToken);
```
[the remark this answers is not in the document]
Comment (mikethebeer) [reply enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: addressed here https://github.com/crate/crate/pull/6519/files#diff-45a9155fc2f450871ac368a058b044edR180

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java [hunk 7]
```diff
@@ -30,87 +32,159 @@
...
+                                }
+                                hash = parser.binaryValue();
+                                break;
+                            case "salt":
+                                if (currentToken == XContentParser.Token.VALUE_NULL) {
+                                    throw new ElasticsearchParseException(
+                                        "failed to parse SecureHash, 'salt' value is not a binary value [{}]", currentToken);
```
[the remark this answers is not in the document]
Comment (mikethebeer) [reply enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: https://github.com/crate/crate/pull/6519/files#diff-45a9155fc2f450871ac368a058b044edR180

File: enterprise/users/src/test/java/io/crate/metadata/SecureHashTest.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        SecureHash hash1 = SecureHash.of(PASSWORD, rand1);
+        SecureRandom rand2 = SecureRandom.getInstance("SHA1PRNG");
+        rand2.setSeed(1);
+        SecureHash hash2 = SecureHash.of(PASSWORD, rand2);
+        assertEquals(hash1, hash2);
+    }
+
+    @Test
+    public void testSamePasswordsGenerateDifferentHash() throws Exception {
+        SecureRandom rand = new SecureRandom();
```
[the remark this answers is not in the document]
Comment (mikethebeer) [reply enterprise/users/src/test/java/io/crate/metadata/SecureHashTest.java]: using `getInstance()` as discussed: https://github.com/crate/crate/pull/6519/files#diff-45a9155fc2f450871ac368a058b044edR72

File: enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java (hunk 4 above)
Comment (mikethebeer) [reply enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: hm. difficult since the generated hash is not predictable / deterministic.

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java (hunk 3 above)
Comment (mikethebeer) [reply enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: @mxm not sure if I get you right here. However I think it's fine if we apply the migration if any cluster state change occurs.

File: enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java (hunk 4 above)
Comment (seut) [reply enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: I agree with @mxm, we should verify here that the user can execute a query with the defined password.

Comment (chaudum) [reply enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: there will be other tests with the last PR
https://github.com/crate/crate/commit/d71db3731be03f8a1037f21271173ec27364f117#diff-0293163e51773d21ddef26a112a9eaa2

File: enterprise/users/src/test/java/io/crate/metadata/SecureHashTest.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        assertTrue(hash.verifyHash(PASSWORD));
+        assertFalse(hash.verifyHash(INVALID_PASSWORD));
+    }
+
+    @Test
+    public void testNonAsciiChars() throws InvalidKeySpecException, NoSuchAlgorithmException {
+        SecureString pw = new SecureString("πä😉ـص".toCharArray());
+        SecureHash hash = SecureHash.of(pw);
+
+        assertFalse(hash.verifyHash(INVALID_PASSWORD));
```
Comment (seut) [inline enterprise/users/src/test/java/io/crate/metadata/SecureHashTest.java]: this seems to be irrelevant for this test

File: enterprise/users/src/test/java/io/crate/operation/user/TransportUserActionTest.java [hunk 10]
```diff
@@ -76,18 +79,22 @@ public void testDropUserNoUsersAtAll() throws Exception {
...
+            new UsersMetaData(UserDefinitions.SINGLE_USER_ONLY),
             "trillian"
         );
         assertThat(res, is(false));
     }
 
     @Test
     public void testDropUser() throws Exception {
-        UsersMetaData oldMetaData = new UsersMetaData(ImmutableList.of("ford", "arthur"));
+        Map<String, SecureHash> users = new HashMap<>();
```
Comment (seut) [inline enterprise/users/src/test/java/io/crate/operation/user/TransportUserActionTest.java]: why not use DUMMY_USERS here?

File: enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java (hunk 3 above)
Comment (mxm) [reply enterprise/users/src/main/java/io/crate/metadata/UsersMetaData.java]: It would be nice to have just one way of reading user data and migrating older formats of the user data with the migration service. @chaudum explained it's not really possible to perform the migration before reading any user meta data. So I guess we have to leave it as it is.

File: enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java (hunk 4 above)
Comment (mxm) [reply enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: @mikethebeer For a full integration test, testing if you can authenticate with the newly created user would be nice.

File: sql/src/main/java/io/crate/analyze/CreateUserAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        this.functions = functions;
+    }
+
+    public CreateUserAnalyzedStatement analyze(CreateUser node, ParamTypeHints typeHints, TransactionContext txnContext) {
+        Map<String, Expression> properties = node.properties().properties();
+        ExpressionAnalyzer expressionAnalyzer = new ExpressionAnalyzer(
+            functions,
+            txnContext,
+            typeHints,
+            new FieldProvider() {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/CreateUserAnalyzer.java]: can be a static instance

File: sql/src/main/java/io/crate/analyze/symbol/ParamSymbols.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze.symbol;
+
+import io.crate.data.Row;
+
+public class ParamSymbols extends DefaultTraversalSymbolVisitor<Row, Symbol> {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/symbol/ParamSymbols.java]: pls make it final and add a private constructor to prevent mis-usage

File: sql/src/main/java/io/crate/user/UserActions.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+import static io.crate.concurrent.CompletableFutures.failedFuture;
+
+public final class UserActions {
+
+    private UserActions() {
+
+    }
+
+    public static CompletableFuture<Long> executeCreateUser(BiFunction<String, SecureHash, CompletableFuture<Long>> userAction,
```
Comment (seut) [inline sql/src/main/java/io/crate/user/UserActions.java]: sry but still don't think the name is reflecting what this method is all about. the main thing here done is resolving the password and generating a hash out of it (and yes then applies the given BiFunction). maybe just let it generate the hash and do the execute out of this function like it was before? also this would be more correct at the exception handling as afaik the catched exception can only occur while generating the hash and not while applying the bifunction. by doing this, the related test method would also be much simpler and clear

File: enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java (hunk 4 above)
Comment (mikethebeer) [reply enterprise/users/src/test/java/io/crate/integrationtests/UserManagementIntegrationTest.java]: @seut as @chaudum said this will be possible with the last PR. This PR only enhances the statement and the storage/migration of user-meta data.

File: sql/src/main/java/io/crate/analyze/CreateUserAnalyzer.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                public Symbol resolveField(QualifiedName qualifiedName, @Nullable List path, Operation operation) {
+                    throw new UnsupportedOperationException("Can only resolve string literals");
+                }
+            },
+            null
+        );
+
+        Map<String, Symbol> rows = new HashMap<>();
+        ExpressionAnalysisContext exprCtx = new ExpressionAnalysisContext();
+        for (String key : properties.keySet()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateUserAnalyzer.java]: iteration over the entryset would be more efficient - as it avoids the get lookup per key
